File: scala/src/main/scala/datastructures/DynamicArray.scala
package datastructures

import scala.collection.mutable.ArrayBuffer

/**
 * Resizable array with automatic capacity management.
 * Time: O(1) amortized for push, O(n) for insert/remove
 * Space: O(n)
 */
class DynamicArray[T]:
  private val items = ArrayBuffer.empty[T]

  def push(value: T): this.type =
    items += value
    this

  def pop(): Option[T] =
    if items.isEmpty then None
    else Some(items.remove(items.length - 1))

  def get(index: Int): Option[T] =
    if index < 0 || index >= items.length then None
    else Some(items(index))

  def set(index: Int, value: T): Boolean =
    if index < 0 || index >= items.length then false
    else
      items(index) = value
      true

  def insert(index: Int, value: T): Boolean =
    if index < 0 || index > items.length then false
    else
      items.insert(index, value)
      true

  def removeAt(index: Int): Option[T] =
    if index < 0 || index >= items.length then None
    else Some(items.remove(index))

  def indexOf(value: T): Int = items.indexOf(value)

  def contains(value: T): Boolean = items.contains(value)

  def isEmpty: Boolean = items.isEmpty

  def size: Int = items.length

  def clear(): this.type =
    items.clear()
    this

  def toList: List[T] = items.toList

File: kotlin/src/main/kotlin/datastructures/DynamicArray.kt
package datastructures

/**
 * Resizable array with automatic capacity management.
 * Time: O(1) amortized for add, O(n) for insert/remove
 * Space: O(n)
 */
class DynamicArray<T> {
    private var items: Array<Any?> = arrayOfNulls(DEFAULT_CAPACITY)
    private var count = 0

    companion object {
        private const val DEFAULT_CAPACITY = 16
    }

    fun push(value: T) {
        ensureCapacity(count + 1)
        items[count++] = value
    }

    fun pop(): T? {
        if (count == 0) return null
        @Suppress("UNCHECKED_CAST")
        val value = items[--count] as T
        items[count] = null
        return value
    }

    @Suppress("UNCHECKED_CAST")
    fun get(index: Int): T? {
        if (index < 0 || index >= count) return null
        return items[index] as T
    }

    fun set(index: Int, value: T): Boolean {
        if (index < 0 || index >= count) return false
        items[index] = value
        return true
    }

    fun insert(index: Int, value: T): Boolean {
        if (index < 0 || index > count) return false
        ensureCapacity(count + 1)
        for (i in count downTo index + 1) {
            items[i] = items[i - 1]
        }
        items[index] = value
        count++
        return true
    }

    fun removeAt(index: Int): T? {
        if (index < 0 || index >= count) return null
        @Suppress("UNCHECKED_CAST")
        val value = items[index] as T
        for (i in index until count - 1) {
            items[i] = items[i + 1]
        }
        items[--count] = null
        return value
    }

    fun indexOf(value: T): Int {
        for (i in 0 until count) {
            if (items[i] == value) return i
        }
        return -1
    }

    fun contains(value: T): Boolean = indexOf(value) != -1

    fun isEmpty(): Boolean = count == 0

    fun size(): Int = count

    fun clear() {
        for (i in 0 until count) {
            items[i] = null
        }
        count = 0
    }

    @Suppress("UNCHECKED_CAST")
    fun toList(): List<T> = (0 until count).map { items[it] as T }

    private fun ensureCapacity(minCapacity: Int) {
        if (minCapacity > items.size) {
            val newCapacity = maxOf(items.size * 2, minCapacity)
            items = items.copyOf(newCapacity)
        }
    }
}
